ts/main.ts

```diff
@@ -409,3 +466,108 @@ function clearGrid(): void {
     }
   }
 }
+
+const $switcher = document.querySelector('.switcher-square') as HTMLDivElement;
+if (!$switcher) throw new Error('.switcher-square query failed!');
+const $categoryModal = document.querySelector(
+  '.grid-cats',
+) as HTMLDialogElement;
+if (!$categoryModal) throw new Error('.grid-cats query failed!');
+const $gridSetsContainer = document.querySelector(
+  '.grid-sets-container',
+) as HTMLDivElement;
+if (!$gridContainer) throw new Error('.grid-sets-container query failed!');
+
+$switcher.addEventListener('click', function (): void {
+  $categoryModal.show();
+});
+
+for (const categorySet in categorySets) {
+  const gridRow = document.createElement('div');
+  gridRow.classList.add(
+    'flex',
+    'py-1',
+    'ml-2',
+    'rounded-xl',
+    'hover:bg-yellow-200',
+    'main-content',
+    'select-grid',
+  );
+  gridRow.setAttribute('data-grid-ID', categorySet);
+  if (currentGridId === parseInt(categorySet)) {
+    gridRow.classList.add('bg-yellow-500', 'selected-grid-cat-row');
+    gridRow.classList.remove('hover:bg-yellow-200');
+  }
+
+  const imgContainer = document.createElement('div');
+  imgContainer.classList.add('w-1/12', 'main-content');
+  const puzzImg = document.createElement('img');
+  puzzImg.setAttribute('src', './images/puzzle.png');
+  puzzImg.classList.add('main-content');
+  imgContainer.appendChild(puzzImg);
+
+  const gridRowTxt = document.createElement('h1');
+  gridRowTxt.textContent = `Grid #${gridRow.getAttribute('data-grid-ID')}`;
+  gridRowTxt.classList.add('main-content');
+
+  gridRow.appendChild(imgContainer);
+  gridRow.appendChild(gridRowTxt);
+
+  $gridSetsContainer.appendChild(gridRow);
+}
+
+$categoryModal.addEventListener('click', function (event: Event): void {
+  const element = event.target as HTMLElement;
+  const closestCatRow = element.closest('.select-grid') as HTMLElement;
+
+  const $selectedRow = document.querySelector(
+    '.selected-grid-cat-row',
+  ) as HTMLDivElement;
+  if (!$selectedRow) throw new Error('.selected-grid-cat-row query failed!');
+
+  if (
+    !element.classList.contains('main-content') ||
+    closestCatRow.classList.contains('selected-grid-cat-row')
+  ) {
+    $categoryModal.close();
+  } else if (
+    element.tagName === 'IMG' ||
+    element.tagName === 'H1' ||
+    element.tagName === 'DIV'
+  ) {
```

Comment:
Is there something else we can check in here rather than checking three different elements? You used `closest` in your `if` condition but what is this checking for?

Reply:
it's set up like this cause i want to close the modal if the user clicks anything off of the main content or selects the row of the grid categories that's already selected. Otherwise, if they click any part of a row(row itself(DIV), puzzle icon(IMG), or text labeling the grid ID(H1)) that is associated with a category set that is not the one already selected I want to execute the block of code that switches the categories

Reply:
I use closest because whether I click on the row itself(div), icon(img), or text(h1) I need to get the row element's data-grid-ID attribute value to pass into the newGridCategories() function as an argument to render the appropriate new grid categories that the user wants to select

Reply:
each row(div) has the 'select-row' class so whether I click on the row itself(div), puzzle icon(img), or text(h1) i have guaranteed access to the data-grid-ID attribute that I need to change to the categories by passing it in to the newGridCategories() function